### ISSUE TITLE
Predict Work Item Type, Classification, and Support Group

### ISSUE THREAD

**Issue by AdhocAdam** [AdhocAdam/smletsexchangeconnector#112]
Integrate the solution seen on the Azure Machine Learning gallery [here
](https://gallery.azure.ai/Experiment/Email-Classification-for-Automated-Support-Ticket-Generation-Step-1-of-2-Train-and-Evaluate-Models-3) by leveraging one's own SCSM data warehouse to train models and deploy a web service that the connector could consume in order to predict the Work Item type (Incident or Service Request), Support Group, and Classification.

**Comment by AdhocAdam** [comment]
To do:

- [ ] Blog post/wiki/some kind of walkthrough on how to deploy and configure Azure ML for the connector
- [ ] Create the function, variables, and means of integration to the web service
- [ ] Ensure correct placement with respect to the Dynamic Analyst Assignment feature #111 so as to provide complimentary feature set

**Comment by AdhocAdam** [comment]
Once AML is trained and the web service published, the following function can be used to return a Probability object to use within the connector. While this function requires only the API key, AML url, and an email subject/body - the actual implementation will use ACS style functionality wherein some minimum percent is used before the value is set.

```powershell
#config variables
$amlAPIKey = ""
$amlSite = ""

function Get-AMLWorkItemProbability ($EmailSubject, $EmailBody)
{
    #create the header
    $headerTable = @{"Authorization" = "Bearer $amlAPIKey"; "Content-Type" = "application/json"}

    #create the JSON request
    $messagePayload = @"
    {
        "Inputs": {
            "Input1" : {
                "ColumnNames": ["Email_Subject", "Email_Description"],
                "Values": [
                    ["$EmailSubject", "$EmailBody"]
                ]
            }
        },
        "GlobalParameters": {}
    }
"@

    #invoke the Azure Machine Learning web service for predicting Work Item Type, Classification, and Support Group
    $probabilityResponse = Invoke-RestMethod -Uri $amlSite -Method Post -Header $headerTable -Body $messagePayload -ContentType "application/json"

    #return custom probability object
    $probabilityResults = $probabilityResponse.Results.output1.value.Values[0]
    $probabilityMatrix = New-Object -TypeName psobject
    $probabilityMatrix | Add-Member -MemberType NoteProperty -Name WorkItemType -Value $probabilityResults[0]
    $probabilityMatrix | Add-Member -MemberType NoteProperty -Name WorkItemTypeConfidence -Value $probabilityResults[1]
    $probabilityMatrix | Add-Member -MemberType NoteProperty -Name WorkItemClassification -Value $probabilityResults[2]
    $probabilityMatrix | Add-Member -MemberType NoteProperty -Name WorkItemClassificationConfidence -Value $probabilityResults[3]
    $probabilityMatrix | Add-Member -MemberType NoteProperty -Name WorkItemSupportGroup -Value $probabilityResults[4]
    $probabilityMatrix | Add-Member -MemberType NoteProperty -Name WorkItemSupportGroupConfidence -Value $probabilityResults[5]

    #return the percent score
    return ($probabilityMatrix)
}

Get-AMLWorkItemProbability -EmailSubject "my account" -EmailBody "I changed my password recently, but now I can't seem to log in. Can you reset it again?"
```

**Comment by AdhocAdam** [comment]
SQL query to generate the CSV dataset along the lines of:

```SQL
SELECT ir.Title as 'Email_Subject',
ir.Description as 'Email_Description',
ir.Classification as 'Case_Subject',
ir.TierQueue as 'Queue_Name',
'1' as 'Case_CaseType'
FROM IncidentDim as ir
WHERE ir.Source like '%email%'
AND ir.createddate > '1-1-2019'

UNION ALL

SELECT sr.Title as 'Email_Subject',
sr.Description as 'Email_Description',
sr.Area as 'Case_Subject',
sr.SupportGroup as 'Queue_Name',
'0' as 'Case_CaseType'
FROM ServiceRequestDim as sr
WHERE sr.Source like '%email%'
AND sr.createddate > '1-1-2019'
```

**Comment by AdhocAdam** [comment]
Core integration with the connector is complete and can be leveraged alongside Azure Cognitive Services for Sentiment Analysis. Reviewing the Pull.

Still proofing Wiki walkthrough (e.g. sql queries for compatibility with returned data, missing screenshots, etc.) for configuring AML as well as the AML Experiments to be published to the gallery.

**Comment by AdhocAdam** [comment]
Experiments published to the gallery.

- https://gallery.azure.ai/Experiment/01-SCSM-Email-Work-Item-Classification-Train-and-Evaluate
- https://gallery.azure.ai/Experiment/02-SCSM-Email-Work-Item-Classification-Create-Web-Service-w-Trained-Models

**Comment by AdhocAdam** [comment]
Outside of some possible grammatical errors, the general flow of the Wiki setup is done and highlights things such as:

- Free/Standard accounts, their configuration, and pricing
- SQL queries to generate the dataset for ML to train against and example of modification from SCSM DW
- PowerShell function taken directly from the connector to test functionality outside of the connector
- Origin of project (Shashi Lanka Venkata)
- Origin of dataset (Microsoft/Endava) as well as how it was modified

**Comment by AdhocAdam** [comment]
To be released on v1.4.7/v1.5 of the connector (March 2019 release)
